imap/conversations.c

```diff
@@ -285,22 +285,56 @@ EXPORTED int conversations_num_folders(struct conversations_state *state)
     return strarray_size(state->folders);
 }
 
-EXPORTED const char* conversations_folder_name(struct conversations_state *state,
-                                               int foldernum)
+EXPORTED const char *conversations_folder_mboxname(const struct conversations_state *state,
+                                                   int foldernum)
 {
-    const char *folder = strarray_safenth(state->folders, foldernum);
-    if (state->folders_byname) {
-        return folder;
-    }
-    else {
-        static char fname[MAX_MAILBOX_NAME+1];
+    const char *val = strarray_nth(state->folders, foldernum);
+    if (!strcmpsafe(val, "-")) return NULL;  // tombstone
+
+    if (state->folders_byname)
+        return val;
+
+    // make it writeable even though const, because we're fiddling cache
+    struct conversations_state *backdoor = (struct conversations_state *)state;
+    if (!backdoor->altrep) backdoor->altrep = strarray_new();
+    const char *res = strarray_nth(backdoor->altrep, foldernum);
+    if (!res) {
+        if (!val) return NULL;
+        // cache lookup
         mbentry_t *mbentry = NULL;
+        int r = mboxlist_lookup_by_uniqueid(val, &mbentry, NULL);
+        if (r || !mbentry) return NULL;
+        strarray_set(backdoor->altrep, foldernum, mbentry->name);
+        mboxlist_entry_free(&mbentry);
+        res = strarray_nth(backdoor->altrep, foldernum);
+    }
+    return res;
+}
 
-        mboxlist_lookup_by_uniqueid(folder, &mbentry, NULL);
-        strlcpy(fname, mbentry ? mbentry->name : "", sizeof(fname));
```

Comment:
This method seems like it always used to return a string or an empty string, never NULL (unless ` const char *folder = strarray_safenth(state->folders, foldernum);` could have been null?).

This new version returns NULL which you've checked for in one place, but there are others that need checking too in imap/jmap_mail.c just above it I think.

Should it instead return "" like it used to or do we always need to guard the responses from this with NULL checks?

Reply:
I'd prefer to guard with NULL checks - because "" is clearly not a thing.

Reply:
I have done a commit that puts those checks in!